Allow ignoring soft wraps when moving to line ends

tverghis opened the pull request:
Release Notes:

- Fixed #10888

This patch addresses behavior of `Editor::move_to_{beginning|end}_of_line`. It adds a setting, `stop_at_soft_wraps` when defining a keymap for the `editor::MoveToBeginningOfLine` and `editor::MoveToEndOfLine` actions. When `true`, it causes movement to the either end of the line (via, for example Home or End), to go to the logical end, as opposed to the nearest soft wrap point in the respective direction.

tverghis committed:
Add stop_at_soft_wraps for begin/end of line move

This allows configuring whether moving to the beginning or end of a line
(for example, with the Home/End) keys should stop at soft wraps or not.

By default, this is false -- i.e., movement to the beginning or end of a
line will move to the logical position, rather than the position
dictated by the nearest soft wrap.

tverghis committed:
Fix move to line beginning/end behavior

Respect the `stop_at_soft_wraps` field for both
`Editor::move_to_beginning_of_line` and `Editor::move_to_end_of_line`.

This changes the default behavior: previously, this would set the cursor
at the nearest soft wrap location in the respective direction. Now, it
will set the cursor properly at the logical beginning/end by default.

tverghis commented:
@cla-bot check

tverghis committed:
Fix struct usage in tests

tverghis commented:
Looks like this broke some assertions in `editor_tests::test_beginning_end_of_line`. I'll try to fix that, and add some additional tests for when `stop_at_soft_wraps` is true (old behavior).

SomeoneToIgnore commented on crates/editor/src/actions.rs:
```diff
@@ -106,10 +118,12 @@ impl_actions!(
         SelectNext,
         SelectPrevious,
         SelectToBeginningOfLine,
+        MoveToBeginningOfLine,
```
Let's sort these alphabetically.

tverghis replied:
Done

SomeoneToIgnore commented on crates/editor/src/actions.rs:
```diff
@@ -13,6 +13,12 @@ pub struct SelectPrevious {
 
+#[derive(PartialEq, Clone, Deserialize, Default)]
+pub struct MoveToBeginningOfLine {
+    #[serde(default)]
```
This changes the current default, which corresponded to `stop_at_soft_wraps: true`, let's keep it unchanged instead and use https://github.com/zed-industries/zed/blob/8a79535b84dfee2f88a3ad865465893ec3cdf540/crates/language/src/language_settings.rs#L421 or similar idea for that.

tverghis replied:
Is the current behavior (i.e., stopping at soft-wraps) desired, or is it a bug? Defaulting to `false` would be a change to existing behavior, but maybe that is fine if the behavior is unintended?

If the current behavior is intended, then this PR becomes more of an enhancement ("allow configuring stop-at-soft-wrap when moving to line ends") than a bug fix. In that case I should update the PR title and release notes. I don't mind either way -- let me know what you think!

SomeoneToIgnore replied:
Yeah, I'd say this looks more like an improvement and the old behavior is intended (at least, for some group of people).

SomeoneToIgnore reviewed (commented):
Looks nice, let's add a small test, similar to what `test_beginning_end_of_line` has (or just another few lines straight there)?

tverghis committed:
Refactor default_true to module

tverghis committed:
Default true for stop_at_soft_wraps

Previously, this setting would default to `false` when navigating to
line ends. Now, it will default to true.

tverghis committed:
Sort impl_actions section

tverghis committed:
Add tests

tverghis commented:
@SomeoneToIgnore I believe my PR is ready for re-review now. Thanks for taking a look!

SomeoneToIgnore reviewed (commented):
Thank you, the changes look complete now — let's fix the tests and have it merged.

tverghis commented:
Sorry for the delay on this.

Unfortunately, it seems that some tests have different results based on whether it runs on CI or on my machine. The tests that are affected by my changes pass on my machine. FWIW, I am on Ubuntu.
```bash
failures:
    display_map::block_map::tests::test_blocks_on_wrapped_lines
    display_map::tests::test_chunks_with_soft_wrapping
    editor_tests::test_add_selection_above_below
    editor_tests::test_beginning_end_of_line
    editor_tests::test_move_cursor_multibyte
    editor_tests::test_prev_next_word_bounds_with_soft_wrap
    hover_links::tests::test_hover_type_links

test result: FAILED. 183 passed; 7 failed; 0 ignored; 0 measured; 0 filtered out; finished in 149.87s
```

It _appears_ that many of the failures likely have something to do with wrapping. Is text wrapping dependent on the machine or OS it runs on?

Anyway, to fix the problem, I will update the expected values to try and make CI happy.

tverghis committed:
Make CI tests happy

tverghis commented:
@SomeoneToIgnore tests have been updated

SomeoneToIgnore commented:
Oh, we do not run tests on Ubuntu yet and I've just stumbled onto similar issues yesterday.
Interesting to see how few tests are left to fix, and sorry for the bad experience on your side — our Linux support is still in the experimental phase.

Thank you for altering those nonetheless, I think it's fine to use mac values for now and fix the Linux tests later.

SomeoneToIgnore commented:
There seems to be another failure, I'll look into this and fix it today.
Hopefully you've ticked the "allow edits from maintainers" in your PR, or you'll have to apply a patch from me 🙂

SomeoneToIgnore commented:
So, checked `map.max_point()` in the `fn line_end` and effectively failed the test on CI, and it returns `DisplayPoint(2, 5)` whilst on your system, that was `DisplayPoint(2, 6)` presumably.

Given the particular set of tests failing, I wonder if that's an off-by-one or other bug in the way we treat display width somewhere in the depth.
We initialize the tests with the same screen width on both platforms, so it's not the test infra code.
I will try to look at it, as interested in fixing Linux tests soon, but no promises yet.

The end diff is just
```
diff --git a/crates/editor/src/editor_tests.rs b/crates/editor/src/editor_tests.rs
index de2161b91..17f2a3c36 100644
--- a/crates/editor/src/editor_tests.rs
+++ b/crates/editor/src/editor_tests.rs
@@ -1288,14 +1288,14 @@ fn test_beginning_end_of_line_ignore_soft_wrap(cx: &mut TestAppContext) {
         // next display line.
         view.move_to_end_of_line(&move_to_end, cx);
         assert_eq!(
-            vec![DisplayPoint::new(2, 6)..DisplayPoint::new(2, 6),],
+            vec![DisplayPoint::new(2, 5)..DisplayPoint::new(2, 5),],
             view.selections.display_ranges(cx)
         );
 
         // Moving to the end of the line again should be a no-op.
         view.move_to_end_of_line(&move_to_end, cx);
         assert_eq!(
-            vec![DisplayPoint::new(2, 6)..DisplayPoint::new(2, 6),],
+            vec![DisplayPoint::new(2, 5)..DisplayPoint::new(2, 5),],
             view.selections.display_ranges(cx)
         );
     });
```

SomeoneToIgnore committed:
Make tests work on mac CI

tverghis commented:
@SomeoneToIgnore thanks so much for patching this up and getting it merged!
